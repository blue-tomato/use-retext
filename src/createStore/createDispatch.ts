import get from 'just-safe-get';
import set from 'just-safe-set';
import mitt from 'mitt';
import { mapValuesDeep } from '../helpers';
import { Reducer } from '../Store';

interface ReducerInfo {
  scope: string;
  reducer: Reducer;
}

export default (initialState: {}, actions: {}): [any, mitt.Emitter] => {
  const groupedReducer: Record<string, Array<ReducerInfo>> = {};
  const emitter = mitt();
  let currentState = initialState;

  return [
    mapValuesDeep(actions, ({ key, value, scope: actionScope }) => {
      if (!groupedReducer[key]) {
        groupedReducer[key] = [];
      }

      groupedReducer[key].push({
        scope: actionScope,
        reducer: (value as Function)(),
      });

      return (payload?: unknown) => {
        groupedReducer[key].forEach(({ scope, reducer }: ReducerInfo) => {
          // Ensure that the scope starts at least with the action-scope
          if (!scope.startsWith(actionScope)) return;

          const state = scope ? get(currentState, scope) : currentState;
          const newState = reducer(state, payload);

          if (scope) {
            set(currentState, scope, {
              ...get(currentState, scope),
              ...newState,
            });
          } else {
            currentState = {
              ...currentState,
              ...newState,
            };
          }
        });

        emitter.emit('update', {
          newState: currentState,
          scope: actionScope,
        });

        // return currentState;
      };
    }),
    emitter,
  ];
};
