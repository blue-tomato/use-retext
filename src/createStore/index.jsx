import React, { createContext, useContext } from 'react';
import useRetext from '../useRetext';

export default store => {
  const Context = createContext();
  const Store = ({ ...props }) => {
    const [state, dispatch] = useRetext(store);

    return <Context.Provider {...props} value={[state, dispatch]} />;
  };
  const useStore = () => useContext(Context);

  return [Store, useStore];
};
