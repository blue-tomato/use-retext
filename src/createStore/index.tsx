import React, { createContext, useContext } from 'react';
import useRetext from '../useRetext';
import { Store } from '../Store';

export default <S extends unknown>(store: Store<S>) => {
  const Context = createContext({});
  const StoreProvider = ({ ...props }) => {
    const { state, dispatch } = useRetext(store);

    return <Context.Provider {...props} value={{ state, dispatch }} />;
  };
  const useStore = () => useContext(Context);

  return [StoreProvider, useStore];
};
