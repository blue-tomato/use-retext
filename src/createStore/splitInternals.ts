import reduce from 'just-reduce-object';
import typeOf from 'just-typeof';
import { Store } from '../Store';

const splitInternals = (store: Store) =>
  reduce(
    store,
    (acc: Array<Record<string, unknown>>, key: string, value: any) => {
      const state = acc[0];
      let actions = acc[1];

      if (value.__action) {
        // Add to actions
        actions = {
          ...actions,
          [key]: value.__action,
        };
      } else if (typeOf(value) === 'object') {
        // Extract inner store
        const [extractedState, extractedActions] = splitInternals(value as Store);

        state[key] = extractedState;

        // Don't add empty objects
        if (Object.keys(extractedActions).length >= 1) {
          actions[key] = extractedActions;
        }
      } else {
        // Add to state
        state[key] = value;
      }

      return [state, actions];
    },
    [{}, {}],
  );

export default splitInternals;
