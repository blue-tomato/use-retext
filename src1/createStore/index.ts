import useRetext, { SelectorFunction } from '../useRetext';
import { Store } from '../Store';

export default (store: Store) => {
  const useStore = (selectorFunction?: SelectorFunction) => {
    if (!selectorFunction) {
      return useRetext(store);
    }

    const selectedStore = selectorFunction(store);

    return useRetext(selectedStore);
  };

  return useStore;
};
