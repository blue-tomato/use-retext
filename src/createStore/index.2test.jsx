import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { renderHook, act } from '@testing-library/react-hooks';
import createStore from '.';

const { store } = global;

describe('return', () => {
  it('returns a store and useStore', () => {
    const [Store, useStore] = createStore(store);
    const wrapper = shallow(<Store />);

    expect(useStore).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('in application usage', () => {
  it('updates the state', () => {
    const [Store, useStore] = createStore(store);
    const { result } = renderHook(() => useStore(store), { wrapper: Store });

    act(() => {
      result.current.dispatch.increment();
      result.current.dispatch.sideMenu.increment();
      result.current.dispatch.sideMenu.toggle();
    });

    expect(result.current.state).toMatchSnapshot();
  });
});
