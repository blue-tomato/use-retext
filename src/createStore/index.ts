import { useEffect, useMemo, useState } from 'react';
import get from 'just-safe-get';
import { Reducer, Store } from '../Store';
import splitInternals from './splitInternals';
import createDispatch from './createDispatch';

interface UpdateInfo {
  scope: string;
  newState: object;
}

const action = (reducer?: Reducer) => ({
  // Create a function so it won't get mapped like an object
  __action: () => reducer,
});

export default (store: Store) => {
  const [initialState, actions] = splitInternals(store);
  const [dispatch, emitter] = createDispatch(initialState, actions);

  return (path?: string) => {
    const [state, setState] = useState(() => (path ? get(initialState, path) : initialState));
    const selectedDispatch = useMemo(() => (path ? get(dispatch, path) : dispatch), [path]);

    useEffect(() => {
      const stateUpdater = ({ newState, scope }: UpdateInfo) => {
        if (path && scope.startsWith(path)) {
          setState(get(newState, path));
        } else {
          setState(newState);
        }
      };

      emitter.on('update', stateUpdater);

      return () => emitter.off('update', stateUpdater);
    }, [path]);

    return { state, dispatch: selectedDispatch };
  };
};
export { action };
