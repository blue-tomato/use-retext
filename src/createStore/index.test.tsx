import { FunctionComponent } from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { renderHook, act } from '@testing-library/react-hooks';
import createStore from '.';

const { store } = global;

describe('return', () => {
  it('returns a store and useStore', () => {
    const [Store, useStore] = createStore(store);

    expect(useStore).toMatchSnapshot();
    expect(Store).toMatchSnapshot();
  });
});

describe('in application usage', () => {
  it('updates the state', () => {
    const [Store, useStore] = createStore(store);
    const { result } = renderHook(() => useStore(store), { wrapper: Store as FunctionComponent });

    act(() => {
      result.current.dispatch.increment();
      result.current.dispatch.sideMenu.increment();
      result.current.dispatch.sideMenu.toggle();
    });

    expect(result.current.state).toMatchSnapshot();
  });
});
