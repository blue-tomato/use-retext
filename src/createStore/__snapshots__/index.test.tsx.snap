// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`return returns a store and useStore 1`] = `[Function]`;

exports[`return returns a store and useStore 2`] = `[Function]`;
